### PR TITLE
Fix ownership issues around ConnectionPool and evhtp_connection_t.

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: cpp/net/connection_pool.cc [hunk 1]
```diff
@@ -81,14 +82,82 @@ static Gauge<string>* connections_per_host_port(
                        "Number of cached connections port host:port"));
 
 
+namespace {
+
+
+// static
```
Comment (pphaneuf) [inline cpp/net/connection_pool.cc]: Static?

Comment (AlCutter) [inline cpp/net/connection_pool.cc]: C&P'd, removed.

File: cpp/net/connection_pool.cc [hunk 2]
```diff
@@ -354,22 +413,38 @@ unique_ptr<ConnectionPool::Connection> ConnectionPool::Get(const URL& url) {
...
   if (it == conns_.end() || it->second.empty()) {
     VLOG(1) << "new evhtp_connection for " << key.first << ":" << key.second;
-    unique_ptr<ConnectionPool::Connection> conn(new Connection(
+    // This EvConnection has a slightly complicated lifetime; it needs to hang
+    // around until libevhtp/libevent have entirely finished with the
+    // evhtp_connection_t it references, and for at least as long as the life
+    // of the Connection we return from this method.
+    //
+    // This is accomplished through the use of a couple of shared_ptrs;
+    // this one, which goes inside the returnec Connection object, and another
```
Comment (pphaneuf) [inline cpp/net/connection_pool.cc]: s/returnec/returned/

Thanks for this comment!

File: cpp/net/connection_pool.cc [hunk 3]
```diff
@@ -307,18 +353,31 @@ evhtp_res ConnectionPool::Connection::ConnectionErrorHook(
...
 
 // static
+evhtp_res EvConnection::ConnectionFinishedHook(evhtp_connection_t* conn,
+                                               void* arg) {
+  CHECK_NOTNULL(conn);
+  CHECK_NOTNULL(arg);
+  CHECK(libevent::Base::OnEventThread());
+  // libevhtp has finished with this connection so we can release the
+  // shared_ptr we had to keep it around until now.
+  unique_ptr<shared_ptr<EvConnection>> const c(
```
Comment (pphaneuf) [inline cpp/net/connection_pool.cc]: Yo dawg, I herd you like smart pointers... :wink:

File: cpp/net/connection_pool.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/net/connection_pool.cc]: Fixed

File: cpp/net/connection_pool.cc (hunk 3 above)
Comment (AlCutter) [inline cpp/net/connection_pool.cc]: :)

**Comment by pphaneuf** [comment]
LGTM

I expect that by 2030, a reasonable C++ HTTP/HTTPS client library will exist.

**Comment by AlCutter** [comment]
Hahaha, you hopeless optimist.

**Comment by pphaneuf** [comment]
Just in time for when it'll be phased out for HTTP2 or something... :stuck_out_tongue:

**Pull request merged**